flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -70,17 +80,30 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(Fabric8FlinkKubeClient.class);
 
+    /**
+     * Cache for Flink kubernetes clients. Key is master URL and namespace. Value is
+     * Fabric8FlinkKubeClient and reference count.
+     */
+    @GuardedBy("LOCK")
+    private static final Map<Tuple2<String, String>, Tuple2<Fabric8FlinkKubeClient, Integer>>
+            CACHE = new HashMap<>();
+    /** Lock used to protect modification of CACHE. */
+    private static final Object LOCK = new Object();
```

Comment:
Do we really need this complexity? What do we gain by it? We save instantiating 4 * x threads, right?